src/ICS20Transfer.sol

```diff
@@ -72,59 +67,49 @@ contract ICS20Transfer is IIBCApp, IICS20Transfer, IICS20Errors, Ownable, Reentr
             revert ICS20MsgSenderIsNotPacketSender(msg_.sender, packetData.sender);
         }
 
-        _transferFrom(packetData.sender, address(this), packetData.erc20ContractAddress, packetData.amount);
+        // transfer the tokens to us (requires the allowance to be set)
+        _transferFrom(packetData.sender, address(this), packetData.erc20Contract, packetData.amount);
+
+        // if the denom is prefixed by the port and channel on which we are sending
+        // the token, then we must be returning the token back to the chain they originated from
+        if (packetData.receiverChainIsSource) {
+            // receiver chain is source: burn the vouchers
+            // TODO: Implement escrow balance tracking (#6)
+            IBCERC20 ibcERC20Contract = IBCERC20(packetData.erc20Contract);
```

Comment:
Wondering if we should check the balance before and after the transfer before the burn. Let's say if we have a to deal with a malicious implementation of _transferFrom this may protect us. But it's more a thought future audits.

Reply:
The `_transferFrom` that happens before the burn has that check already, actually, and the burn itself is safe as it only happens in our own ERC20 contract.

```solidity
function _transferFrom(address sender, address receiver, address tokenContract, uint256 amount) private {
    // we snapshot our current balance of this token
    uint256 ourStartingBalance = IERC20(tokenContract).balanceOf(receiver);

    IERC20(tokenContract).safeTransferFrom(sender, receiver, amount);

    // check what this particular ERC20 implementation actually gave us, since it doesn't
    // have to be at all related to the _amount
    uint256 actualEndingBalance = IERC20(tokenContract).balanceOf(receiver);

    uint256 expectedEndingBalance = ourStartingBalance + amount;
    // a very strange ERC20 may trigger this condition, if we didn't have this we would
    // underflow, so it's mostly just an error message printer
    if (actualEndingBalance <= ourStartingBalance || actualEndingBalance != expectedEndingBalance) {
        revert ICS20UnexpectedERC20Balance(expectedEndingBalance, actualEndingBalance);
    }
}
```